### ISSUE TITLE
Fmt testing

### ISSUE THREAD

**Issue by PXshadow** [go2hx/go2hx#153]
# Tests:
* [x] TestErrorf
* [x] TestFmtInterface
* [x] TestSprintf
* [x] TestComplexFormatting
* [x] TestReorder
* [x] TestCountMallocs
* [x] TestFlagParser
* [x] TestStructPrinter
* [x] TestSlicePrinter
* [x] TestMapPrinter
* [x] TestEmptyMap
* [x] TestBlank
* [x] TestBlankln
* [x] TestFormatterPrintln
* [x] TestWidthAndPrecision
* [x] TestPanics
* [x] TestBadVerbRecursion
* [x] TestIsSpace
* [x] TestNilDoesNotBecomeTyped
* [x] TestFormatterFlags
* [x] TestParsenum
* [x] TestAppendf
* [x] TestAppend
* [x] TestAppendln
* [x] TestScan
* [x] TestScanln
* [ ] TestScanf
* [ ] TestScanOverflow
* [ ] TestNaN
* [ ] TestInf
* [ ] TestScanfMulti
* [ ] TestScanMultiple
* [ ] TestScanEmpty
* [ ] TestScanNotPointer
* [ ] TestScanlnNoNewline
* [ ] TestScanlnWithMiddleNewline
* [ ] TestEOF
* [ ] TestEOFAtEndOfInput
* [ ] TestEOFAllTypes
* [ ] TestUnreadRuneWithBufio
* [ ] TestMultiLine
* [ ] TestLineByLineFscanf
* [ ] TestScanStateCount
* [ ] TestScanInts
* [ ] TestHexBytes
* [ ] TestScanNewlinesAreSpaces
* [ ] TestScanlnNewlinesTerminate
* [ ] TestScanfNewlineMatchFormat
* [ ] TestHexByte
* [ ] TestStringer
# Outline
Progress will be documented in the comments below and this list will be edited accordingly. The hopes of doing it in this way is to allow easy contribution of one of the most pivotal Golang pkg to get passing (and to help with my sanity a little).
# Completion goals
Allow go2hx to finally start compiling simple libraries, begin progress for an early alpha release of go2hx and start stabilizing the compiler and building a solid base for use of the compiler.

**Comment by PXshadow** [comment]
# TestSprintf
``ptr`` kind when calling ``reflect.Value.Bool()`` method causes panic expected ``bool`` kind.
```go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	tt := struct {
		fmt string
		val any
		out string
	}{"%s", &[3]byte{'a', 'b', 'c'}, "&abc"}
	s := fmt.Sprintf(tt.fmt, tt.val)
	i := strings.Index(tt.out, "PTR")
	if i >= 0 && i < len(s) {
		var pattern, chars string
		switch {
		case strings.HasPrefix(tt.out[i:], "PTR_b"):
			pattern = "PTR_b"
			chars = "01"
		case strings.HasPrefix(tt.out[i:], "PTR_o"):
			pattern = "PTR_o"
			chars = "01234567"
		case strings.HasPrefix(tt.out[i:], "PTR_d"):
			pattern = "PTR_d"
			chars = "0123456789"
		case strings.HasPrefix(tt.out[i:], "PTR_x"):
			pattern = "PTR_x"
			chars = "0123456789abcdef"
		case strings.HasPrefix(tt.out[i:], "PTR_X"):
			pattern = "PTR_X"
			chars = "0123456789ABCDEF"
		default:
			pattern = "PTR"
			chars = "0123456789abcdefABCDEF"
		}
		p := s[:i] + pattern
		for j := i; j < len(s); j++ {
			if !strings.ContainsRune(chars, rune(s[j])) {
				p += s[j:]
				break
			}
		}
		s = p
	}
	if s != tt.out {
		if _, ok := tt.val.(string); ok {
			// Don't requote the already-quoted strings.
			// It's too confusing to read the errors.
			fmt.Println(fmt.Errorf("Sprintf(%q, %q) = <%s> want <%s>", tt.fmt, tt.val, s, tt.out))
			os.Exit(0)
		} else {
			fmt.Println(fmt.Errorf("Sprintf(%q, %v) = %q want %q", tt.fmt, tt.val, s, tt.out))
			os.Exit(0)
		}
	}
}
```

**Comment by elliott5** [comment]
`reflect.Value.Bool()` is behaving correctly, see https://go.dev/play/p/E-wfS9F0Sve .
So the issue must be with the calling code, if the tests have been written correctly.

**Comment by PXshadow** [comment]
Latest go version (go1.20rc3) updates fmt test to be more expansive and add more cases to testErrorf, and new test code to handle each one of the cases.

Therefore now no tests are passing and the current issue is:
```
Can't cast stdgo.errors.T_errorString_asInterface to stdgo.SliceData
```

From fmt_test splitErr function.

**Comment by elliott5** [comment]
From the Go language spec [on copy special handling](https://go.dev/ref/spec#Appending_and_copying_slices)

> The function copy copies slice elements from a source src to a destination dst and returns the number of elements copied. The [core types](https://go.dev/ref/spec#Core_types) of both arguments must be slices with [identical](https://go.dev/ref/spec#Type_identity) element type. The number of elements copied is the minimum of len(src) and len(dst). As a special case, if the destination's core type is []byte, copy also accepts a source argument with core type [bytestring](https://go.dev/ref/spec#Core_types). This form copies the bytes from the byte slice or string into the byte slice.
> ```go
> copy(dst, src []T) int
> copy(dst []byte, src string) int
> ```

Which demonstrates that sometimes strings can auto-convert to []byte in Go.

But in this case you would expect `errorString` to require a function call in order to get a string that could be converted.

This may be related to the error in the `interp test` error in https://github.com/go2hx/go2hx/blob/master/stdgo/fmt/README.md , presumably from the previous version of Go:
```
Exception: Unexpected value VFunction(#fun), expected instance
```
whereas the equivalent `hl test` error was similar to what you reported:
```
Exception: Can't cast stdgo.reflect.Value_asInterface to stdgo.reflect.Value
```

**Comment by PXshadow** [comment]
The latest commit fmt tests are passing up until TestSprintf, fmtTests number 505. An issue occurs with asInterface casting of reflect Value to reflect Value_asInterface, the next commit I came up with the solution of simply not allowing asInterface casting for reflect Value.

**Comment by PXshadow** [comment]
TestSprintf, fmtTests number 506
```go
Sprintf("%#v", {1 2 a [1 2]}) = "A{i:1, j:0x2, s:\"a\", x:[]int{1, 2}}" want "fmt_test.A{i:1, j:0x2, s:\"a\", x:[]int{1, 2}}"
```

The go version includes the ``fmt_test.`` suffix where as the go2hx version does not.

**Comment by elliott5** [comment]
> TestSprintf, fmtTests number 506
> 
> ```go
> Sprintf("%#v", {1 2 a [1 2]}) = "A{i:1, j:0x2, s:\"a\", x:[]int{1, 2}}" want "fmt_test.A{i:1, j:0x2, s:\"a\", x:[]int{1, 2}}"
> ```
> 
> The go version includes the `fmt_test.` suffix where as the go2hx version does not.

This must be the go2hx implementation of `reflect.Type.String()`:
```go
	// String returns a string representation of the type.
	// The string representation may use shortened package names
	// (e.g., base64 instead of "encoding/base64") and is not
	// guaranteed to be unique among types. To test for type identity,
	// compare the Types directly.
	String() string
```
which should, according to the spec above, be returning `fmt_test.A` rather than just `A`.

**Comment by PXshadow** [comment]
TestSprintf, fmtTests number 512:
```go
Sprintf("%#v", map[]) = "map[string]int{}" want "map[string]int{\"a\":1}"
```
map string type not outputting key values.

**Comment by PXshadow** [comment]
TestSprintf, fmtTests number 618,619,620,621,622: 

```haxe
Exception: reflect.Value.Bytes
Called from stdgo.fmt.$T_pp_static_extension._printValue (stdgo/fmt/Fmt.hx line 3091)
```

when reflect.bytes for Value is called an exception occurs.

**Comment by PXshadow** [comment]
TestSprintf, fmtTest number 625:

```haxe
Exception: reflect.complex is not yet implemented
```

reflect Value.complex is not yet implemented.

**Comment by PXshadow** [comment]
Interesting to note there is a total of 768 tests for fmtTests which only testSprintf uses.

**Comment by elliott5** [comment]
> TestSprintf, fmtTests number 512:
> 
> ```go
> Sprintf("%#v", map[]) = "map[string]int{}" want "map[string]int{\"a\":1}"
> ```
> 
> map string type not outputting key values.

It is not outputting either the keys or the values - so not iterating over the contents of the map.

**Comment by PXshadow** [comment]
@elliott5 Yes the value is not actually being set in the map, here is the problem isolated.

```go
package main

import (
	"fmt"
)

type I int

func (i I) String() string { return fmt.Sprintf("<%d>", int(i)) }

type B struct {
	I I
	j int
}

func main() {
	x := map[string]B{"a": {1, 2}}
	for key, value := range x {
		fmt.Println(key, value)
	}
}
```

**Comment by elliott5** [comment]
That's not the same issue as https://github.com/go2hx/go2hx/issues/153#issuecomment-1420278839 @PXshadow , which would be:
```go
package main

import (
	"fmt"
)

func main() {
	println(fmt.Sprintf("%#v", map[string]int{"a": 1}))
}
```

**Comment by elliott5** [comment]
Which is not to say that outputting a map containing a named type might not be a separate problem...

**Comment by PXshadow** [comment]
> 

This now works, though I'm not sure if the changes have all been committed yet.

**Comment by PXshadow** [comment]
> That's not the same issue as [#153 (comment)](https://github.com/go2hx/go2hx/issues/153#issuecomment-1420278839) @PXshadow , which would be:
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> )
> 
> func main() {
> 	println(fmt.Sprintf("%#v", map[string]int{"a": 1}))
> }
> ```

This now works though I'm not sure if the required changes are committed yet.

**Comment by PXshadow** [comment]
**Sprintf tests are now passing!!**

So the new trench line is as follows:

TestStructPrinter
```go
Sprintf("%#v", s) = `fmt_test.testStructPrinter_0{a:"abc", b:"def", c:123}`, want `fmt_test.T{a:"abc", b:"def", c:123}`
```

**Comment by elliott5** [comment]
This is great news @PXshadow , well done!

And the new "trench line" issue  "only" looks like getting the correct "Go" name printed out, when it is from a local type declared inside a function. This is a wonderfully obscure bug!

**Comment by PXshadow** [comment]
Indeed I think so too @elliott5 ! I should also note that that there is a small regression for only the sort tests (strings, and bytes has been fixed now) so I will get that resolved shortly, but it is also part of the front line or trench line haha.

For the local type inside of a function issue, I think I will add a suffix to the name so that go2hx knows to resolve it to the short go compatible name when it is printed out in a string form.

**Comment by elliott5** [comment]
That would by my solution too @PXshadow .

**Comment by PXshadow** [comment]
Solution is now implemented for the suffix naming system, I made a commit for it but wrote the wrong issue number in the message.

**Comment by PXshadow** [comment]
I am skipping TestPanics, as I don't think replicating the exact same internal runtime error messages is useful or worth the time. I spent a decent amount of time making it so non go2hx errors are now supported by the try catch defer system and caught a small bug so the tests were useful to get to the point to see only a fail of the test from different outputs.

**Comment by PXshadow** [comment]
New front lines is now at TestScanf and the first test of _scanfTests:
```sh
Exception: Null access
Called from stdgo.internal.reflect._Reflect.$Reflect_Fields_.directlyAssignable (stdgo/internal/reflect/Reflect.hx line 199)
Called from stdgo.internal.reflect.$_Type_static_extension.assignableTo (stdgo/internal/reflect/Reflect.hx line 1010)
Called from stdgo.internal.reflect._Reflect.$Reflect_Fields_.implementsMethod (stdgo/internal/reflect/Reflect.hx line 377)
Called from stdgo.internal.reflect.$_Type_static_extension.implements_ (stdgo/internal/reflect/Reflect.hx line 1021)
Called from stdgo.internal.reflect.$_Type_static_extension.assignableTo (stdgo/internal/reflect/Reflect.hx line 1010)
```

**Comment by PXshadow** [comment]
Some progress on scannerf here is a minimal issue of the latest problem:

```
func main() {
	var boolValue bool
	fmt.Sscanf("TRUE\n", "%v", &boolValue)
	fmt.Println(boolValue)
}
```

Expected true, but false for go2hx.

**Comment by PXshadow** [comment]
I figured out one issue with it, firstly, is that a switch value can get referenced over and over again. So if the value is a call expr, and the result changes each time it runs, then it messes up the if checks that represent go switches.

For example:
```go
switch x() {
case 0:

case 1:
}
```
turns into approximately:
```haxe
if (x() == 0) {

}else if(x() == 1) {

}
```

Therefore the switch value needs to start being stored in a temporary variable.

**Comment by PXshadow** [comment]
"Variables without an initialization expression do not appear in this list."  -> therefore I can't get order of evaluation from golang to tell me the order of variables where I would've created new values such as Slice and Map if nil with the nilBool enabled.

This kills my current plan of creating nilBool for slice and map to get around the issue of a pointer not being ``nil`` but ``slice`` or ``map`` being ``nil``. My previous approach is flawed because it's just a simple type holder, only really used by macros, it has no means to hold anything else because it's meant to be light weight as pass by ref values already acts like pointers.

So I may have to create a third version of Pointer, NullRef where NullRef can not be null but the value pointing can be.

Here's the two side by side.

```haxe
typedef Ref<T> = T;
class NullRef<T> {

}

class NullRef<T> {
    public var value:T;
    public function new(value) {
        this.value = value;
    }
}
```

I will most likely wrap NullRef before in an abstract to make it just as easy to use as the ``Ref`` type.

I'm curious on your thoughts @elliott5 .

**Comment by elliott5** [comment]
My first reaction is that "Ref<T>" should be an enum with two states - "refNull" or "refValue<T>".

...but there is probably more to it than that!

**Comment by elliott5** [comment]
Example code to illustrate the idea is [here](https://try.haxe.org/#B1B4eA00) @PXshadow .

...still don't know if it goes some way to answering the question you were asking.

**Comment by PXshadow** [comment]
Thanks for the example code @elliott5 I'll try a similar approach to this soon and report back here how it worked.

**Comment by elliott5** [comment]
You are right, the example does not work because a Pointer type can only work on a Haxe value that is passed by reference - doh!

**Comment by PXshadow** [comment]
Haha no problem. I really want to keep ```typedef Ref<T> = T``` it's really clean and allows the code to be really neat, if this changes then it effects a lot more things then I realized, it makes it harder to access all pass as ref pointer values which is very common in Haxe to be able to be used seamlessly, it breaks the static extension system, and requires the same hacky solution of passing multiple versions of its self as a static extension a non ref version first followed by a ref arg into the function.

Soo back to the drawing board here are my current ideas:

* go static analysis to mark nil pointers that are pass as ref values and have the compiler set at compile time whether the value is nil or not when an equality expr occurs. I know this can't cover all cases.

* have it so pass as ref pointer always returns false, this will run into issues where a ref equality check results in trying to use a nil map or slice.

* Use an enum or a class for ref, this will make the code really hard to handle for Haxe devs, but the performance could be minimized with the ``inline new`` class system that would turn the initialization into variables. This might be the only solution to have this problem work in all cases, but I find it's a really high cost to pay for something we might be able to mostly cover all of the use cases.

What do you think @elliott5 ?

**Comment by PXshadow** [comment]
The 2nd idea is not a great idea, what's better is the original approach which is just checking the underlying Ref value if that is nil for whether Ref is nil.

```haxe
var v any
var p *any = &v
println(p == nil)
```

Haxe reports true, Go reports false, all other types of cases work the same (p is not nil, if v is nil)

**Comment by elliott5** [comment]
The issue here is how to initialise empty variables.

The Go spec on [variables](https://go.dev/ref/spec#Variables) says:
> If a variable has not yet been assigned a value, its value is the [zero value](https://go.dev/ref/spec#The_zero_value) for its type.

This means that an un-initialised variable (like `var v` above) should never be `nil` (unless it is a pointer). Rather it should be an empty version of its type.

If this were done in the example above, `v` would be an empty copy of a Go `any` emulation (held by reference), and, like Go, Haxe would also report false for `p == nil`.

This would also mean that all variables containing [Haxe Basic Types](https://haxe.org/manual/types-basic-types.html) that could be pointed to would need to be held by reference, presumably using anonymous structures.

**Comment by PXshadow** [comment]
TestScanf number 85:
```go
package main

import (
	"fmt"
)

// Xs accepts any non-empty run of the verb character
type Xs string

func (x *Xs) Scan(state fmt.ScanState, verb rune) error {
	tok, err := state.Token(true, func(r rune) bool { return r == verb })
	if err != nil {
		return err
	}
	s := string(tok)
	/*if !regexp.MustCompile("^" + string(verb) + "+$").MatchString(s) {
		return errors.New("syntax error for xs")
	}*/
	*x = Xs(s)
	return nil
}

var xVal Xs

func main() {
	fmt.Sscanf("  sss ", "%s", &xVal)
	fmt.Println(xVal)
}
```

```sh
Uncaught exception: Null access
Called from stdgo.fmt.$T_ss_static_extension._scanOne(stdgo/fmt/Fmt.hx:4018)
Called from stdgo.fmt.$T_ss_static_extension._doScanf(stdgo/fmt/Fmt.hx:3821)
Called from stdgo.fmt._Fmt.$Fmt_Fields_.fscanf(stdgo/fmt/Fmt.hx:1536)
Called from stdgo.fmt._Fmt.$Fmt_Fields_.sscanf(stdgo/fmt/Fmt.hx:1489)
Called from github_com.go2hx.go4hx.rnd._Rnd.$Rnd_Fields_.main(github_com/go2hx/go4hx/rnd/Rnd.hx:26)
Called from .init(?:1)
```

**Comment by elliott5** [comment]
I can't see the body code for [`_scanOne`](https://github.com/go2hx/go2hx/search?q=_scanOne) is it ever set to a value?

**Comment by PXshadow** [comment]
The above issue is now working, and is now getting blocked by not yet implemented regexp functions.
```sh
Exception: regexp.mustCompile is not yet implemented
```

The next decision is whether to try and patch regexp with Haxe's standard library at the moment or try compiling the pkg or a mixture of both? I'm interested in your thoughts @elliott5 .

**Comment by elliott5** [comment]
The chances of the Go `regexp` system *exactly* matching the equivalent Haxe system is zero.

Also making the Go `regexp` system available to Haxe programmers as an alternative would give them choice.

So I think the only sensible way forward is to compile the `regexp` package @PXshadow .

**Comment by PXshadow** [comment]
@elliott5 makes sense I agree, I'll begin working on ``regexp``'s test suite and make a tracker issue for it.

**Comment by PXshadow** [comment]
Latest:
```
=== RUN  TestErrorf
Exception: unknown type equals: sliceType({get : function#7F953CCA5AB8})
Called from stdgo._AnyInterface.$AnyInterface_Impl_.equals (stdgo/AnyInterface.hx line 183)
Called from stdgo._internal.reflect._Reflect_deepEqual.$Reflect_deepEqual_Fields_.deepEqual (stdgo/_internal/reflect/Reflect_deepEqual.hx line 4)
Called from _internal.fmt_test._Fmt_test_testErrorf.$Fmt_test_testErrorf_Fields_.testErrorf (_internal/fmt_test/Fmt_test_testErrorf.hx line 32)
Called from stdgo._internal.testing.$M_static_extension.run (stdgo/_internal/testing/Testing_M_static_extension.hx line 32)
Called from _internal.fmt_dot_test._Fmt_dot_test.$Fmt_dot_test_Fields_.main (_internal/fmt_dot_test/Fmt_dot_test.hx line 4)
```

**Comment by elliott5** [comment]
```
Sprintf("%#v", s) = `_internal.fmt_test.T_47832{a:"abc", b:"def", c:123}`, want `fmt_test.T{a:"abc", b:"def", c:123}
```

There is probably a need to keep a mapping from `go2hx` generated type names to `go` type names. This could be done as part of the `go reflect` data for a `go` type.

**Comment by PXshadow** [comment]
Other stdlibs to focus on first. Another issue can be made if needed at a later time.
